File: PyProject/leetcode/n-ary-tree-level-order-traversal.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class NAryTreeLevelOrderTraversal:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        q, ret = [root], []
        while any(q):
            ret.append([node.val for node in q])
            q = [child for node in q for child in node.children if child]
        return ret

    def levelOrder2(self, root: 'Node') -> List[List[int]]:
        # 当做队列
        q = [root]
        res = []
        while q:
            temp = []
            next_stack = []
            for node in q:
                temp.append(node.val)
                for child in node.children:
                    next_stack.append(child)
            res.append(temp)
            q = next_stack
        return res

File: GoProject/leetcode/tree/n-ary-tree-level-order-traversal.go
package tree

func levelOrder(root *Node) (res [][]int) {
	var levelorder func(root []*Node)
	levelorder = func(roots []*Node) {
		var nodes []*Node
		var nodeVals []int

		if len(roots) == 0 {
			return
		}
		for _, node := range roots {
			if node != nil {
				nodeVals = append(nodeVals, node.Val)
				for _, n := range node.Children {
					if n != nil {
						nodes = append(nodes, n)
					}
				}
			}
		}
		if len(nodeVals) != 0 {
			res = append(res, nodeVals)
		}
		levelorder(nodes)
	}

	levelorder([]*Node{root})
	return
}
